### ISSUE TITLE
Support multi-threaded downloads

### ISSUE THREAD

**Issue by prateekmedia** [dart-tools/dio_http#3]
### New Issue Checklist

- [x] I have searched for a similar issue in the [project](https://github.com/dart-tools/dio_http/issues) and found none

### Issue Info

| Info                           | Value                                                 |      |
| ------------------------------ | ----------------------------------------------------- | ---- |
| Platform Name                  | flutter                          |      |
| Platform Version               | any                              |      |
| Dio Version                    | latest                                  |      |
| Android Studio / Xcode Version | e.g. Android Studio 3.3.2 / Xcode 10.2.1              |      |
| Repro rate                     | e.g. all the time (100%) / sometimes x% / only once   |      |
| Repro with our demo prj        | e.g. does it happen with our demo project?            |      |
| Demo project link              | e.g. link to a demo project that highlights the issue |      |

### Issue Description and Steps

There were many stale issues on dio's issue tracker regarding this as far as I remember. Multi-threaded downloads are always good for a better downloading experience.

**Comment by jgoyvaerts** [comment]
Do you have some sample code of what's not working correctly? When you say multithreaded, do you mean multiple async/awaits, or Isolates, or something else?

**Comment by prateekmedia** [comment]
First of all everything is working without multi-threading and `dio` and `dio-http` uses single-threading.

What Multi-Threading download does:
- Using multi-threading a file can be downloaded in the form of chunks simultaneously from different threads or different starting positions
- which tries to utilize maximum power of your internet connection. (This will download fast in multi-core processor case, which is explained below).

Multi-threading is what is used in Internet downloader manager or Xtreme download manager or any typical download manager, this is what seperates them from the built-in download manager's

AFAIK no one has used multi threading in dart, but I can reference code from other languages:
- [aria2](https://github.com/aria2/aria2) - C++
- [zou](https://github.com/k0pernicus/zou) - Rust
- [simple download manager](https://www.geeksforgeeks.org/simple-multithreaded-download-manager-in-python/) - Python

This will help to boost downloading speed drastically.

**Comment by jgoyvaerts** [comment]
I'm not sure if that's in scope of the dio_http library. Right now it's meant as a general purpose http library for Dart/Flutter which supports many use cases, and making downloads faster by doing some multithreading or other tricks is a very specific case.

The other libraries you linked are all purely focused on downloading files, they have none of the other features that Dio has.

**Comment by prateekmedia** [comment]
Yes @jgoyvaerts, but maybe in future if you consider it as I don't think there is any other good pure dart downloading library.

`dio`/`dio_http` also lack pause, resume support for now which is also required for downloading big files but that will be for another issue.

**Comment by jgoyvaerts** [comment]
This package is now archived. This was originally forked because the owner of the [Dio](https://github.com/flutterchina/dio) package was inactive, but he has sinced returned from inactiveness and has added some more contributors to prevent this from happening again in the future. All future development continues over on [Dio](https://github.com/flutterchina/dio).
